apisix/plugins/splunk-hec-logging.lua

```diff
@@ -0,0 +1,189 @@
+                },
+                timeout = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 60
+                }
+            },
+            required = { "uri", "token" }
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        },
+        max_retry_count = {
+            type = "integer",
+            minimum = 0,
+            default = 0
+        },
+        retry_delay = {
+            type = "integer",
+            minimum = 0,
+            default = 1
+        },
+        buffer_duration = {
+            type = "integer",
+            minimum = 1,
+            default = 60
+        },
+        inactive_timeout = {
+            type = "integer",
+            minimum = 1,
+            default = 10
+        },
+        batch_max_size = {
+            type = "integer",
+            minimum = 1,
+            default = 100
+        },
+    },
+    required = { "endpoint" },
+}
+
+
+local function get_logger_entry(conf)
+    local entry = log_util.get_full_log(ngx, conf)
+    ngx_update_time()
+    return {
+        time = ngx_now(),
+        host = entry.server.hostname,
+        source = DEFAULT_SPLUNK_HEC_ENTRY_SOURCE,
+        sourcetype = DEFAULT_SPLUNK_HEC_ENTRY_TYPE,
+        event = {
+            request_url = entry.request.url,
+            request_method = entry.request.method,
+            request_headers = entry.request.headers,
+            request_query = entry.request.querystring,
+            request_size = entry.request.size,
+            response_headers = entry.response.headers,
+            response_status = entry.response.status,
+            response_size = entry.response.size,
+            latency = entry.latency,
+            upstream = entry.upstream,
+        }
+    }
+end
+
+
+local function send_to_splunk(conf, entries)
+    if type(conf.endpoint) ~= "table" then
+        return nil, "endpoint config invalid"
+    end
+
+    if not conf.endpoint.uri then
+        return nil, "endpoint url undefined"
+    end
+
+    if not conf.endpoint.token then
+        return nil, "endpoint token undefined"
+    end
+
+    local request_headers = {}
+    request_headers["Content-Type"] = "application/json"
+    request_headers["Authorization"] = "Splunk " .. conf.endpoint.token
+    if conf.endpoint.channel then
+        request_headers["X-Splunk-Request-Channel"] = conf.endpoint.channel
+    end
+
+    local http_new = http.new()
+    local res, err = http_new:request_uri(conf.endpoint.uri, {
+        ssl_verify = conf.ssl_verify,
+        method = "POST",
+        body = core.json.encode(entries),
+        headers = request_headers,
+    })
+
+    if err then
+        return nil, "failed to write log to splunk, " .. err
+    end
+
+    local body
+    body, err = core.json.decode(res.body)
+    if err then
+        return nil, "failed to parse splunk response data, " .. err
+    end
+
+    if res.status ~= 200 then
```

Comment:
If it is a network error `http_new:request_uri`, the process can be terminated after obtaining the err. If it is not for network reasons, the body of the splunk response will be parsed into the table first. I think this is not a problem, because splunk will also respond to the wrong http status Code and errors can only be obtained in the parsed body.

Reply:
The status code might be 500 (unexpected error), or 502 return from the gateway in front of splunk. In these cases, the body may not have `text` field.

Reply:
updated